continuous-deployment/pipelines/recipes/github-workflows/workflow-per-environment.md

```diff
@@ -0,0 +1,31 @@
+# GitHub Workflows
+
+A workflow is a configurable automated process made up of one or more jobs where each of these jobs can be an action in GitHub. Currently a YAML file format is supported for defining a workflow in GitHub.
+
+Additional information on GitHub actions and GitHub Workflows in the links posted in the [references](#References) section below.
+
+## Environments
+
+Environments can be used for different reasons, like some environments are used for development and testing while others are used for production.
+
+Specific environments are protected to prevent unauthorized users from affecting them, since deployments can be executed by different users with different roles. Thus protected environment ensures safety by only letting users with the required privileges, deploy to an environment.
+
+Continuous deployment pipelines help promote changes through automated testing and deployment cycles, out to staging environments and finally to production. The more detailed and comprehensive a testing pipeline is, there is more confidence that the changes wont introduce issues in your production deployment. Since the change must go through the entire process of being tested and deployed, keeping pipelines and deployment environments fast and dependable is of utmost importance.
+
+To understand more about deployment environments, and their differences, please refer to the link in the [references](#References) section below.
+
+## Workflow Per Environment
+
+The general approach is to have one pipeline, where the code is built, tested and deployed and the artifact is then promoted to the next environment, eventually to be deployed into production.
+
+There are multiple ways in GitHub that an environment setup can be achieved. The way it can be done is to have one workflow for multiple environments, but the complexity increases as additional processes and jobs are added to a workflow, which does not mean it cannot be done for small pipelines. The plus point of having one workflow is that, when an artifact flows from one environment to another the state and environment values between the deployment environments can be passed easily.
+
+One way to get around the complexity of a single workflow is to have separate workflows for different environments, making sure that the artifacts from each workflow are independent, as well as, the workflow is small enough to debug any issues seen in any of the workflows. In this case, the state and environment values need to be passed from one deployment environment to another. Multiple workflows also helps to keep the deployments to the environments independent thus reducing the time to deploy and find issues earlier than later in the process. Also, since the environments are independent of each other, any failures in deploying to one environment does not block deployments to other environments.
```

Comment:
> ...making sure that the artifacts from each workflow are independent...

Im not sure I agree with this statement. Are you suggesting that binaries and/or container images be rebuilt within each workflow? How are you going to ensure you have the exact same code as you promote to new env's? I think the only way you could do that without potentially getting other commits mixed in is to target commit hash in each workflow.

Reply:
I do worry a little that recommending the creation of separate but very much closely related workflows might become a maintenance issue down the line too.

Reply:
@gfmatthews : Thats totally true and will add it as a tradeoff for having separate pipelines.

Reply:
As we build more pipelines on github actions, would like to revisit this with more samples. However, think its good starting point for now. 
I am still not sure I agree with whats being said here but, understand available feature set in github actions might be influencing the choices made.

Reply:
I see you make an interesting point, and as we see github actions and workflows is evolving everyday, there will be changes coming in the future for better flow of artifacts.